### PR TITLE
Refactor to support renaming tables

### PR DESCRIPTION
**Pull request by stayradiated**
This PR rewrites `mysql-database` in typescript. I threw away the old class, and wrote a bunch of small functions instead. It's backwards compatible though. The big improvement is that you pass in `entitiesDirectory` - which is scanned and all entities are loaded. 

Each entity can now have a `relationships` function and  a `deprecatedNames` array. The latter is used to provide backwards compatibility when renaming entities.

For example, the following entity:

- Refers to the SQL Table: `Followers`
- Is named `db.SnapchatFollower`
- Can also be accessed through `db.Follower`, `db.Followers` or `db.SnapchatFollowers`
- Has a relation defined to the `UserFollower` table

```javascript
/* entities/snapchatFollower.js */

const SnapchatFollower = (sequelize, DataTypes) => {
  return sequelize.define('Followers', {
    createdAt: DataTypes.DATE,
    updatedAt: DataTypes.DATE,
    snapName: DataTypes.STRING
  })
}

SnapchatFollower.deprecatedNames = [
  'Follower',
  'Followers',
  'SnapchatFollowers'
]

SnapchatFollower.relationships = (self, models) => {
  self.hasMany(models.UserFollower, { foreignKey: 'followerId' })
}

export default SnapchatFollower
```

### CODE REVIEW

File: src/buildModels.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  sequelizeInstance: Sequelize,
+) => {
+  const entityPaths = globSync(['*.js', '!*.spec.js'], {
+    cwd: entitiesDirectory,
+    absolute: true,
+  })
+
+  const deprecatedMap = new Map()
+
+  const entities = entityPaths.map((entityPath) => {
```
Comment (brendonjohn) [inline src/buildModels.ts:32]: Let's give this function a name and export it.

Something like `requireEntities` or `importEntities` could do.

This function should have unit tests.

File: src/buildModels.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    const entity = require(entityPath).default
+    if (Object.hasOwnProperty.call(entity, 'deprecatedNames')) {
+      for (const name of entity.deprecatedNames) {
+        deprecatedMap.set(name, entity.name)
+      }
+    }
+    return entity
+  })
+
+  const models = entities.reduce((models, entity) => {
```
Comment (brendonjohn) [inline src/buildModels.ts]: Let's give this function a name and export it.

Something like `buildSequelizeEntities` could do.

This function should have unit tests.

File: src/buildModels.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    const model = sequelizeInstance.import(entity.name, entity)
+    models[entity.name] = model
+    return models
+  }, {})
+
+  const modelProxy = new Proxy(models, {
```
Comment (brendonjohn) [inline src/buildModels.ts]: can we add a test for the proxy thingy.

File: src/buildModels.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        return true
+      }
+      if (deprecatedMap.has(prop)) {
+        return true
+      }
+      return false
+    },
+  })
+
+  entities.forEach((entity) => {
```
Comment (brendonjohn) [inline src/buildModels.ts:65]: Lets move this to its own function and write a unit test for it.

**Review by brendonjohn — CHANGES_REQUESTED** [review]
I love where this is going. I'm really happy with it.

Building the entities and relationships needs to be split up and have tests written.

**Comment by stayradiated** [comment]
[ch20771] https://app.clubhouse.io/mishguru/story/20771/mysql-database-data-mishmeta-refactor-to-support-renaming-tables

**Comment by stayradiated** [comment]
Thank you for your review @brendonjohn! Totally agree about needing tests and splitting up the `buildModels.js` function.

**Pull request merged**